core/base/src/TObject.cxx

```diff
@@ -751,6 +751,9 @@ void TObject::UseCurrentStyle()
 ///  Using the kWriteDelete option a previous key with the same name is
 ///  deleted only after the new object has been written. This option
 ///  is safer than kOverwrite but it is slower.
+///  NOTE: Neither kOverwrite nor kWriteDelete reduces the size of a TFile--
```

Comment:
Note that Delete/Purge never shrinks the file.   What it does is make the space (usually somewhere in the middle of the file) available for reuse.
Note that the notion/separation of data vs meta-data that is mentioned in this addition is valid **solely** for TTree (and of course its derived clases TNTuple).
For example for an histogram the "data" and "meta-data" are both stored together in a single key and thus the kWriteDelete frees up/reuse all the space occupied for the histograms.

For TTree the story is more complicated.   In the usual case the data is written once (each branches will write its own series of TBaskets which are TKeys so, as far as the file is concerned a 'top level item' that is not registered with any TDirectory).

On the other hand the meta-data is usually written many time.  The TTree object (which contains the meta-data which includes where to find the data in the disk) is usually written many time (to allow for data recovery in case the process crashes before the natural end).  When you write this TTree object, you may want

  * be safe and keep the back copy (cycles) for a while (this is the default, leads to 2 cycles per TTree)
  * be slightly less safe and use kWriteDelete
  * be daring and use kOverwrite

In ***none*** of those cases would you want the data to disappear ... because then it would be gone forever since the TTree object does not contain the data (except for the 'last' baskets) so the only copy is in the file.

When users want to throwaway the data and replace it with something else, they usually have just created new files.

Even if/when we support the "throwaway the data and replace it" mode, it will still be less efficient that creating a new file, as the new data may or may not fit in the 'hole in the file' left by the data-removal.

Reply:
Added line clarifying the distinction between `TTree`s and other objects.